Spring Boot Security 1.0

aksoli666 opened the pull request:
added spring-boot-starter-security
added SecurityConfig
created CustomUserDetailsService
implemented UserDetails in the User
/auth/** -> public, other -> accessible before auth added role entity and repo layer
added all liquibase changesets
used @PreAuthorisation

aksoli666 committed:
Spring Boot Security 1.0
added spring-boot-starter-security
added SecurityConfig
created CustomUserDetailsService
implemented UserDetails in the User
/auth/** -> public, other -> accessible before auth
added role entity and repo layer
added all liquibase changesets
used @PreAuthorisation

aksoli666 committed:
Spring Boot Security 2.0
insert roles to table
But I have a problem sending requests to create a book from a user with the role of user, the request is successful, but it should not work like this. How do I properly fix the code because I can't find the error?

aksoli666 commented:
Hi, I have a problem sending requests to create a book from a user with the role of user, the request is successful, but it should not work like this. How do I properly fix the code because I can't find the error?

liliia-ponomarenko commented on src/main/java/shop/dto/request/UserRegistrationRequestDto.java:
```diff
@@ -26,4 +28,5 @@ public class UserRegistrationRequestDto {
     @NotBlank
     private String lastName;
     private String shippingAddress;
+    private Set<Role> roles;
```
NotEmpty

liliia-ponomarenko commented on src/main/java/shop/entity/User.java:
```diff
@@ -26,4 +39,43 @@ public class User {
     @Column(nullable = false)
     private String lastName;
     private String shippingAddress;
+    @ManyToMany(fetch = FetchType.EAGER)
```
Don't use EAGER fetch type

liliia-ponomarenko commented on src/main/resources/db/changelog/changes/01-create-books-table.yaml:
```diff
@@ -41,7 +41,7 @@ databaseChangeLog:
               - column:
                   name: is_deleted
-                  type: boolean
+                  type: bit
```
```suggestion
                  type: tinyint(1)
```

liliia-ponomarenko commented on src/main/resources/db/changelog/changes/04-create-users-table.yaml:
```diff
@@ -37,4 +37,9 @@ databaseChangeLog:
                   name: shipping_address
                   type: varchar(255)
-
+              - column:
+                  name: is_deleted
+                  type: bit
+                  constraints:
+                    nullable: false
+                    
```
This is normal at the training stage, but you should understand that on real projects you will never change existing files, but always create a new one for new changes

liliia-ponomarenko reviewed (changes requested):
Good job! Let’s fix a few mistakes ;)

aksoli666 committed:
Spring Boot Security 3.0
change liquibase for books and users, (field isDeleted was bit, now tinyint(1))
used entity graph to avoid lie
roles not empty

Elena-Bruyako commented on src/main/java/shop/config/SecurityConfig.java:
```diff
@@ -0,0 +1,53 @@
+import shop.service.user.details.CustomUserDetailsService;
+
+@EnableWebSecurity
+@EnableMethodSecurity
```
leave only one annotation

Elena-Bruyako commented on src/main/java/shop/dto/request/CreateBookRequestDto.java:
```diff
@@ -21,4 +21,5 @@ public class CreateBookRequestDto {
     private BigDecimal price;
     private String description;
     private String coverImage;
+
```
remove empty line

Elena-Bruyako commented on src/main/java/shop/dto/request/UserRegistrationRequestDto.java:
```diff
@@ -26,4 +29,6 @@ public class UserRegistrationRequestDto {
     @NotBlank
     private String lastName;
     private String shippingAddress;
+    @NotEmpty
+    private Set<Role> roles;
```
don't need to provide roles in UserRegistrationRequestDto
```suggestion
```

Elena-Bruyako commented on src/main/java/shop/dto/responce/UserResponseDto.java:
```diff
@@ -11,4 +13,5 @@ public class UserResponseDto {
     private String firstName;
     private String lastName;
     private String shippingAddress;
+    private Set<Role> roles;
```
```suggestion
```

Elena-Bruyako commented on src/main/java/shop/entity/User.java:
```diff
@@ -26,4 +40,45 @@ public class User {
     @Column(nullable = false)
     private String lastName;
     private String shippingAddress;
+    @ManyToMany()
```
```suggestion
    @ManyToMany
```

Elena-Bruyako commented on src/main/java/shop/entity/User.java:
```diff
@@ -26,4 +40,45 @@ public class User {
+
+    @Override
+    public Collection<? extends GrantedAuthority> getAuthorities() {
+        return roles.stream()
+                .map(role -> new SimpleGrantedAuthority("ROLE_" + role.getRole()))
+                .collect(Collectors.toSet());
```
```suggestion
        return roles;
```

Elena-Bruyako commented on src/main/java/shop/mapper/UserMapper.java:
```diff
@@ -1,13 +1,15 @@
 
 @Mapper(config = MapperConfig.class)
 public interface UserMapper {
+    @Mapping(target = "roles", source = "roles")
```
```suggestion
```

Elena-Bruyako commented on src/main/java/shop/service/impl/UserServiceImpl.java:
```diff
@@ -23,6 +25,7 @@ public UserResponseDto register(UserRegistrationRequestDto registerUserRequestDt
             throw new RegistrationException("Can't register user with given email");
         }
         User user = userMapper.toUser(registerUserRequestDto);
+        user.setPassword(passwordEncoder.encode(registerUserRequestDto.getPassword()));
```
set ROLE USER
```suggestion
        user.setPassword(passwordEncoder.encode(registerUserRequestDto.getPassword()));
        user.setRole(Set.of(....
```

Elena-Bruyako reviewed (changes requested):
see comments

aksoli666 committed:
Spring Boot Security 3.0
dіdn't to provide roles in UserRegistrationRequestDto and response.

ahoienko commented on src/main/java/shop/repository/role/RoleRepository.java:
```diff
@@ -0,0 +1,8 @@
+import shop.entity.Role;
+
+public interface RoleRepository extends JpaRepository<Role, Long> {
+    Role findByRole(Role.RoleName role);
```
```suggestion
    Optional<Role> findByRole(Role.RoleName role);
```

ahoienko reviewed (changes requested):
very good, left minor comment

aksoli666 committed:
Spring Boot Security 3.0
findByRole return Optional<Role>

JJJazl commented on src/main/java/shop/repository/role/RoleRepository.java:
```diff
@@ -0,0 +1,9 @@
+import shop.entity.Role;
+
+public interface RoleRepository extends JpaRepository<Role, Long> {
+    Optional<Role> findByRole(Role.RoleName role);
```
```suggestion
    Optional<Role> findByRoleName(Role.RoleName roleName);
```

JJJazl commented on src/main/java/shop/service/impl/UserServiceImpl.java:
```diff
@@ -23,6 +30,9 @@ public UserResponseDto register(UserRegistrationRequestDto registerUserRequestDt
         }
         User user = userMapper.toUser(registerUserRequestDto);
+        user.setPassword(passwordEncoder.encode(registerUserRequestDto.getPassword()));
+        user.setRoles(Set.of(roleRepository.findByRole(Role.RoleName.ROLE_USER)
+                .orElseThrow(() -> new EntityNotFoundException("Can`t find role"))));
```
Let's make the code more explicit and the exception messages more informative
```suggestion
RoleName userRoleName = Role.RoleName.ROLE_USER
        user.setRoles(Set.of(roleRepository.findByRoleName(userRoleName)
                .orElseThrow(() -> new EntityNotFoundException("Can`t find role: " + userRoleName.name()))));
```
